### PR TITLE
Upgrade handshake cipher to AES CBC

### LINKED ISSUE #3360: [Feature]: Upgrade handshake cipher to AES CBC
### Feature Request

The handshake process requires decrypting, stripping a prefix from and reversing the challenge, then re-encrypting it.
The cipher used is "AES" which means "AES/ECB/PKCS5Padding".
"ECB" ciphers are disrecommended in-general because they are vulnerable to replay attacks, chosen plaintext attacks and are easier to build a code-book for.
This means such ciphers aren't always available in cipher suites, which makes interoperability more difficult, and while I know of no vulnerability upgrading may be a useful precaution.

### The Solution you'd like

Change the cipher to "AES/CBC/PKCS5Padding", which involves changing the handshake to include exchanging IV parameters.

### Alternatives that you've considered.

De-scope password support from my hobby client, or find another library that will accept ECB upstream.

### Additional Context

_No response_

### PR DESCRIPTION
**Pull request by fishface60**
### Identify the Bug or Feature request

resolves #3360 

### Description of the Change

Changes the CryptoUtil cipher to "AES/CBC/PKCS5Padding", adds an IV field to handshake messages, and generates a new IV per message.

In the process, found a couple of bugs, and removed CipherUtil objects, which were only used to immediately retrieve the key out of.

### Possible Drawbacks

Slightly more CPU cost involved in generating and transmitting the IV, and chaining blocks together.

It's more changes to cryptography that could result in a vulnerability.

### Release Notes

- Fixed a bug which could rarely cause a false handshake failure
- Fixed error message reporting public key authentication failure while using password authentication.
- Upgraded handshake cipher to CBC for better interoperability and precaution.

---
This change is [Reviewable](https://reviewable.io/reviews/rptools/maptool/3361)

### CODE REVIEW

File: src/main/proto/handshake.proto [hunk 1]
```diff
@@ -40,11 +40,13 @@ message ClientInitMsg {
 message UseAuthTypeMsg {
   AuthTypeEnum auth_type = 1;
   bytes salt = 2;
+  bytes iv = 4;
```
Comment (thelsing) [inline src/main/proto/handshake.proto:43]: any reason what 4 is before 3?

**Review by thelsing — APPROVED** [review]
I'm no crypto expert, but the changes look fine to me. I also think it's better to change from ECB to CBC.

File: src/main/proto/handshake.proto (hunk 1 above)
Comment (fishface60) [reply src/main/proto/handshake.proto:43]: Just that conceptually the salt and IV belong together as encryption parameters and it makes sense for the challenge to come after, but my understanding of the reason why protobufs lets you specify the field order is that they're theoretically backwards compatible.
MapTool doesn't need to support that so I can't think of any problems with changing the order.

**Comment by thelsing** [comment]
Successfully tested.

**Review by Phergus — APPROVED** [review]
:lgtm:

> Reviewed 16 of 16 files at r1, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/rptools/maptool/3361#-:-N-97VddA8UhZUeh0KPy:bnfp4nl)* status: all files reviewed, 1 unresolved discussion (waiting on @thelsing)

**Comment by Phergus** [comment]
@thelsing  anytime you comment on code you must resolve the discussion for the PR to be accepted.

**Review by thelsing — COMMENTED** [review]
> Reviewed 16 of 16 files at r1, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/rptools/maptool/3361)* status: all files reviewed, 1 unresolved discussion (waiting on @thelsing)

**Pull request merged**